File: kotlin/Problem020_ValidParentheses.kt
// https://leetcode.com/problems/valid-parentheses/

class Solution {
    private val parentheses = mapOf('(' to ')', '[' to ']', '{' to '}')


    fun isValid(s: String): Boolean {

        val stack = ArrayDeque<Char>()

        for (c in s) {
            if (c in parentheses.keys) {
                stack.add(c)
            } else {
                val last = stack.removeLastOrNull() ?: return false
                if (parentheses[last] != c) return false
            }
        }

        return stack.isEmpty()
        
    }
}

File: go/problem020_valid_parentheses.go
// https://leetcode.com/problems/valid-parentheses/

func isValid(s string) bool {
    if len(s)%2 !=0 {
            return false
    }
    bracket := map[rune]rune {
        '(':')',
        '{':'}',
        '[':']',
    }

    var stack []rune

    for _, c := range s {

        if closeBracket, ok := bracket[c]; ok {
            stack = append(stack, closeBracket)
        } else {
            if len(stack) == 0 || stack[len(stack)-1] != c {
                return false
            }
            stack = stack[:len(stack)-1]
        }

    }
    return len(stack) == 0    

}

